File: backend/mcp-code-intelligence-nodejs/src/orchestration/local/transport.ts
/**
 * Transport detection and server process interface.
 * Determines whether a server entry should use stdio or httpStream transport.
 */

import { ServerEntry, TransportType } from '../config.js';
import { ServerState } from './process.js';

/** Common interface for both stdio and httpStream server processes. */
export interface IServerProcess {
  readonly name: string;
  state: ServerState;
  tools: Record<string, any>[];
  retryCount: number;
  start(): Promise<boolean>;
  stop(): void;
  restart(maxRetries: number): Promise<boolean>;
  callTool(toolName: string, args: Record<string, any>, timeoutMs: number): Promise<any>;
  healthCheck(): Promise<boolean>;
  isAlive(): boolean;
}

/**
 * Detect transport type from server entry config.
 * - url present, no command → httpStream
 * - command present, no url → stdio
 * - both present → use transportType field (default httpStream)
 * - neither → stdio (will fail at spawn, preserves existing error behavior)
 */
export function detectTransport(entry: ServerEntry): TransportType {
  if (entry.url && !entry.command) return 'httpStream';
  if (entry.command && !entry.url) return 'stdio';
  if (entry.url && entry.command) return entry.transportType === 'stdio' ? 'stdio' : 'httpStream';
  return 'stdio';
}

File: backend/mcp-code-intelligence-python/src/mcp_code_intel/orchestration/local/transport.py
"""Transport detection and server process protocol.

Determines whether a server entry should use stdio or httpStream transport.
Behavioral parity with Kotlin Transport.kt and NodeJS transport.ts.
"""

from __future__ import annotations

from typing import Any, Protocol

from ..config import ServerEntry, TransportType
from .process import ServerState


class IServerProcess(Protocol):
    """Common interface for both stdio and httpStream server processes."""

    name: str
    state: ServerState
    tools: list[dict]
    retry_count: int

    async def start(self) -> bool: ...
    def stop(self) -> None: ...
    async def restart(self, max_retries: int) -> bool: ...
    async def call_tool(self, tool_name: str, args: dict, timeout_ms: int) -> Any: ...
    async def health_check(self) -> bool: ...
    def is_alive(self) -> bool: ...


def detect_transport(entry: ServerEntry) -> TransportType:
    """Detect transport type from server entry config.

    - url present, no command → httpStream
    - command present, no url → stdio
    - both present → use transport_type field (default httpStream)
    - neither → stdio (will fail at spawn, preserves existing error behavior)
    """
    has_url = bool(entry.url)
    has_command = bool(entry.command)
    if has_url and not has_command:
        return TransportType.HTTP_STREAM
    if has_command and not has_url:
        return TransportType.STDIO
    if has_url and has_command:
        return TransportType.STDIO if entry.transport_type == TransportType.STDIO else TransportType.HTTP_STREAM
    return TransportType.STDIO
